src/util/query_exec_util.cpp

```diff
@@ -0,0 +1,283 @@
+  if (!result.first || !result.second) {
+    *success = false;
+    return 0;
+  }
+
+  const std::unique_ptr<planner::AbstractPlanNode> &out_plan = result.second;
+  NOISEPAGE_ASSERT(network::NetworkUtil::DMLQueryType(result.first->GetQueryType()), "ExecuteDML expects DML");
+  common::ManagedPointer<planner::OutputSchema> schema = out_plan->GetOutputSchema();
+
+  auto exec_query = execution::compiler::CompilationContext::Compile(*out_plan, exec_settings_, accessor.get(),
+                                                                     execution::compiler::CompilationMode::OneShot);
+  schemas_.push_back(schema->Copy());
+  exec_queries_.push_back(std::move(exec_query));
+
+  if (require_commit) {
+    EndTransaction(false);
+  }
+
+  *success = true;
+  return exec_queries_.size() - 1;
+}
+
+bool QueryExecUtil::ExecuteQuery(size_t idx, TupleFunction tuple_fn,
+                                 common::ManagedPointer<std::vector<parser::ConstantValueExpression>> params,
+                                 common::ManagedPointer<metrics::MetricsManager> metrics) {
+  NOISEPAGE_ASSERT(idx < exec_queries_.size(), "Invalid query index");
+
+  auto txn = common::ManagedPointer<transaction::TransactionContext>(txn_);
+  bool require_commit = false;
+  if (txn == nullptr) {
+    require_commit = true;
+    BeginTransaction();
+    txn = common::ManagedPointer<transaction::TransactionContext>(txn_);
+  }
+
+  planner::OutputSchema *schema = schemas_[idx].get();
+  auto consumer = [&tuple_fn, schema](byte *tuples, uint32_t num_tuples, uint32_t tuple_size) {
+    if (tuple_fn != nullptr) {
+      for (uint32_t row = 0; row < num_tuples; row++) {
+        uint32_t curr_offset = 0;
+        std::vector<execution::sql::Val *> vals;
+        for (const auto &col : schema->GetColumns()) {
+          auto alignment = execution::sql::ValUtil::GetSqlAlignment(col.GetType());
+          if (!common::MathUtil::IsAligned(curr_offset, alignment)) {
+            curr_offset = static_cast<uint32_t>(common::MathUtil::AlignTo(curr_offset, alignment));
+          }
+
+          auto *val = reinterpret_cast<execution::sql::Val *>(tuples + row * tuple_size + curr_offset);
+          vals.emplace_back(val);
+          curr_offset += execution::sql::ValUtil::GetSqlSize(col.GetType());
+        }
+
+        tuple_fn(vals);
+      }
+    }
+  };
+
+  // Create ExecutionContext with no metrics to prevent recording
+  execution::exec::OutputCallback callback = consumer;
+  auto accessor = catalog_->GetAccessor(txn, db_oid_, DISABLED);
+  auto exec_ctx = std::make_unique<execution::exec::ExecutionContext>(
+      db_oid_, txn, callback, schema, common::ManagedPointer(accessor), exec_settings_, metrics);
+
+  exec_ctx->SetParams(common::ManagedPointer<const std::vector<parser::ConstantValueExpression>>(params.Get()));
+
+  exec_queries_[idx]->Run(common::ManagedPointer(exec_ctx), execution::vm::ExecutionMode::Interpret);
+
+  if (require_commit) {
+    EndTransaction(true);
```

Comment:
Then actually it isn't really `EndTransaction` right, since it isn't called otherwise?

Maybe make `EndTransaction` private, expose a `CommitTransaction` and `AbortTransaction` instead?

Reply:
Restructured this slightly. This is meant to commit any "implicitly" started transactions.